src/Symfony/Component/Encryption/Sodium/SodiumKey.php

```diff
@@ -0,0 +1,163 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Encryption\Sodium;
+
+use Symfony\Component\Encryption\Exception\InvalidKeyException;
+use Symfony\Component\Encryption\KeyInterface;
+
+/**
+ * @internal
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+final class SodiumKey implements KeyInterface
+{
+    /**
+     * @var string|null
+     */
+    private $secret;
+
+    /**
+     * @var string|null
+     */
+    private $privateKey;
+
+    /**
+     * @var string|null
+     */
+    private $publicKey;
+
+    /**
+     * A keypair can only be created from a public and private key.
+     *
+     * @var string|null
+     */
+    private $keypair;
+
+    public static function create(string $secret, string $keypair): self
+    {
+        $key = self::fromSecret($secret);
+        $key->keypair = $keypair;
+        $key->publicKey = sodium_crypto_box_publickey($keypair);
+        $key->privateKey = sodium_crypto_box_secretkey($keypair);
+
+        return $key;
+    }
+
+    /**
+     * The secret key length should be 32 bytes, but other sizes are accepted.
+     */
+    public static function fromSecret(string $secret): self
+    {
+        $key = new self();
+        if (\SODIUM_CRYPTO_SECRETBOX_KEYBYTES === \strlen($secret)) {
+            $key->secret = $secret;
+        } else {
+            // Trim the key to a good size
+            $key->secret = substr(sha1($secret), 0, \SODIUM_CRYPTO_SECRETBOX_KEYBYTES);
+        }
+
+        return $key;
+    }
+
+    public static function fromPrivateKey(string $privateKey): self
+    {
+        $key = new self();
+        $key->privateKey = $privateKey;
+
+        return $key;
+    }
+
+    public static function fromPrivateAndPublicKeys(string $privateKey, string $publicKey): self
+    {
+        $key = new self();
+        $key->privateKey = $privateKey;
+        $key->publicKey = $publicKey;
+
+        return $key;
+    }
+
+    public static function fromPublicKey(string $publicKey): self
+    {
+        $key = new self();
+        $key->publicKey = $publicKey;
+
+        return $key;
+    }
+
+    public static function fromKeypair(string $keypair): self
+    {
+        $key = new self();
+        $key->keypair = $keypair;
```

Comment:
what's about publicKey and privateKey (they are initialized in create method but not here)

Reply:
Yeah. Sure. 

Thank you. This will solve your other comments.